src/main/java/org/openstreetmap/atlas/checks/validation/tag/TollValidationCheck.java

```diff
@@ -0,0 +1,624 @@
+package org.openstreetmap.atlas.checks.validation.tag;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.openstreetmap.atlas.checks.atlas.predicates.TypePredicates;
+import org.openstreetmap.atlas.checks.base.BaseCheck;
+import org.openstreetmap.atlas.checks.flag.CheckFlag;
+import org.openstreetmap.atlas.geography.Heading;
+import org.openstreetmap.atlas.geography.atlas.items.Area;
+import org.openstreetmap.atlas.geography.atlas.items.AtlasObject;
+import org.openstreetmap.atlas.geography.atlas.items.Edge;
+import org.openstreetmap.atlas.geography.atlas.items.Node;
+import org.openstreetmap.atlas.tags.AccessTag;
+import org.openstreetmap.atlas.tags.BarrierTag;
+import org.openstreetmap.atlas.tags.HighwayTag;
+import org.openstreetmap.atlas.tags.TollTag;
+import org.openstreetmap.atlas.utilities.configuration.Configuration;
+import org.openstreetmap.atlas.utilities.scalars.Angle;
+
+/**
+ * This check attempts to validate toll tags based on 3 scenarios. 1. Edge intersects toll feature
+ * but is missing toll tag 2. Edge has inconsistent toll tag compared to surrounding edges 3. Edge
+ * has route that can escape toll feature so the toll tag is modeled incorrectly.
+ *
+ * @author greichenberger
+ */
+public class TollValidationCheck extends BaseCheck<Long>
+{
+    private static final long serialVersionUID = -4286937145318778446L;
+    private static final String INTERSECTS_TOLL_FEATURE = "Way {0, number, #} intersects toll feature but is missing toll tag, please investigate toll tag addition.";
+    private static final String ESCAPABLE_TOLL = "Toll tags need to be investigated for removal on way {0, number, #}. Please check ways {1, number, #} and {2, number, #} and affected nearby ways for modeling issues. Nearby toll features "
+            + "that might be helpful are: upstream {3, number, #} and downstream {4, number, #}.";
+    private static final String INCONSISTENT_TOLL_TAGS = "Way {0, number, #} has an inconsistent toll tag with its surrounding ways. Please check for proper toll tag modeling.";
+    private static final List<String> FALLBACK_INSTRUCTIONS = Arrays.asList(INTERSECTS_TOLL_FEATURE,
+            ESCAPABLE_TOLL, INCONSISTENT_TOLL_TAGS);
+    private static final String HIGHWAY_MINIMUM_DEFAULT = HighwayTag.RESIDENTIAL.toString();
+    private static final Double MAX_ANGLE_DIFF_DEFAULT = 40.0;
+    private static final double MIN_IN_OUT_EDGES_DEFAULT = 1.0;
+    private static final double MAX_ITERATION_FOR_SEARCH_DEFAULT = 15.0;
+    private final Set<Long> markedInconsistentToll = new HashSet<>();
+    private final Set<Long> markedIntersectingTollFeature = new HashSet<>();
```

Comment:
I think we can use BaseCheck's `markAsFlagged` here instead of these Sets, just because it's more conventional and the memory should be allocated already

Reply:
I think I originally was wanting to track these cases individually and not with markAsFlagged but now it should be okay to do so. Will update